### DISCUSSION TITLE
Interceptors

### DISCUSSION THREAD

**Discussion by vibhuh** [axios/axios#6641] [General]
Hi all,
I am developing cloud inventory and accounting application with Qwik JS as frontend. I am using Axios alongwith interceptors for refreshing access tokens whenever the jwt token is expired.

I tried using js-cookie, typescript-cookie for accessing access token which is stored in cookies. The cookie is always undefined as these libraries are not supporting in Qwik js.

Any suggestions can help me a lot.

**Comment by jasonsaayman** [comment]
This is probably a Qwik/runtime issue rather than an Axios interceptor issue.

Qwik can run code on both the server and the browser. Cookie libraries like `js-cookie` only work in the browser because they read `document.cookie`. If your interceptor runs during SSR/server-side execution, `document` does not exist, so the cookie value will be `undefined`.

Also, check whether your cookie is `HttpOnly`. If it is `HttpOnly`, no frontend JavaScript library can read it. That is expected browser security behaviour.

There are two common solutions.

## Option 1: Let the Browser Send the Cookie

If your backend uses cookies for auth, you usually do not need to read the access token in JavaScript. Let the browser send the cookie automatically.

```ts
const api = axios.create({
  baseURL: 'https://api.example.com',
  withCredentials: true
});
```

If your API is on a different origin, your backend must also allow credentials:

```http
Access-Control-Allow-Credentials: true
Access-Control-Allow-Origin: https://your-frontend.example.com
```

Do not use `Access-Control-Allow-Origin: *` with credentials.

Then your refresh interceptor can call the refresh endpoint using the cookie:

```ts
const api = axios.create({
  baseURL: 'https://api.example.com',
  withCredentials: true
});

api.interceptors.response.use(
  response => response,
  async error => {
    const originalRequest = error.config;

    if (
      error.response?.status === 401 &&
      !originalRequest._retry
    ) {
      originalRequest._retry = true;

      await api.post('/auth/refresh');

      return api(originalRequest);
    }

    return Promise.reject(error);
  }
);
```

This works when the refresh token is stored in an `HttpOnly` cookie and the server issues/refreshes auth based on that cookie.

## Option 2: Read Cookies on the Qwik Server

If you are making the API request from Qwik City server-side code, read the cookie from the Qwik request event instead of using `js-cookie`.

Example:

```ts
import { routeLoader$ } from '@builder.io/qwik-city';
import axios from 'axios';

export const useData = routeLoader$(async requestEvent => {
  const accessToken = requestEvent.cookie.get('access_token')?.value;

  const response = await axios.get('https://api.example.com/data', {
    headers: accessToken
      ? {
          Authorization: `Bearer ${accessToken}`
        }
      : undefined
  });

  return response.data;
});
```

That approach works server-side because Qwik has access to the incoming request cookies.

## If You Really Need Browser-Only Cookie Access

Only do this for non-`HttpOnly` cookies, and only in browser-only code:

```ts
if (typeof document !== 'undefined') {
  const token = document.cookie
    .split('; ')
    .find(row => row.startsWith('access_token='))
    ?.split('=')[1];
}
```

But for auth tokens, storing them in JavaScript-readable cookies is usually less secure than using `HttpOnly` cookies.

## Recommendation

For Qwik, the best setup is usually:

- Store refresh/session cookies as `HttpOnly`.
- Use `withCredentials: true` in Axios.
- Let the backend refresh tokens using the cookie.
- If making requests during SSR, read cookies from Qwik City’s `requestEvent.cookie`.
- Avoid a global Axios singleton with per-user auth state on the server, because that can leak auth between requests.
